File: internal/page/page.go
// Package page 定义管道契约中的「分页协议」。
//
// 契约：列表类绑定方法 MUST 返回 PageResult[T]，请求参数 MUST 经 PageRequest.Normalized()
// 归一化后再查库（页码下界、页大小上下界）。
//
// 页大小上下界是【单一真相】：常量定义在本包，service 与护栏共同引用，
// 不得在别处另写一份数字。
package page

// 页大小上下界（单一真相）。
const (
	// DefaultPageSize 请求未指定页大小（0）时的默认值。
	DefaultPageSize = 20
	// MaxPageSize 页大小上限，超出则夹取，防止一次拉全表。
	MaxPageSize = 200
)

// Request 是列表查询的分页请求。
type Request struct {
	// Page 页码，1-based。小于 1 归一化为 1。
	Page int `json:"page"`
	// PageSize 每页条数。0 归一化为 DefaultPageSize，超出上限夹取。
	PageSize int `json:"page_size"`
}

// Normalized 返回归一化后的分页请求：页码下界 1、页大小夹取到 [1, MaxPageSize]。
// service 层在查库前 MUST 调用本方法，避免负数 offset 或一次拉全表。
func (r Request) Normalized() Request {
	out := r
	if out.Page < 1 {
		out.Page = 1
	}
	if out.PageSize <= 0 {
		out.PageSize = DefaultPageSize
	}
	if out.PageSize > MaxPageSize {
		out.PageSize = MaxPageSize
	}
	return out
}

// Offset 返回 SQL offset（归一化后计算）。
func (r Request) Offset() int {
	n := r.Normalized()
	return (n.Page - 1) * n.PageSize
}

// Result 是列表查询的分页结果（泛型）。
//
// 注意（Wails 绑定）：泛型实例在生成的 TS 里会被压平成含模块路径的类名
// （如 PageResult_<pkg>_Item_）。类型安全，名字难看，仅在类型标注处可见。
type Result[T any] struct {
	List     []T   `json:"list"`
	Total    int64 `json:"total"`
	Page     int   `json:"page"`
	PageSize int   `json:"page_size"`
}

// NewResult 构造分页结果，并回填归一化后的 page/page_size，保证前端拿到一致的回显值。
func NewResult[T any](req Request, list []T, total int64) Result[T] {
	n := req.Normalized()
	if list == nil {
		list = []T{} // 保证 JSON 序列化为 []，而非 null，前端无需额外判空
	}
	return Result[T]{
		List:     list,
		Total:    total,
		Page:     n.Page,
		PageSize: n.PageSize,
	}
}

File: frontend/src/lib/page.ts
// 管道契约：Go↔TS 的分页协议前端侧。
//
// 与 Go 侧 internal/page 成对：
//   - 数据结构（Request / Result[T]）在此以 TS 类型镜像，字段名与 JSON tag 一致；
//   - 页大小上下界是【常量镜像】，单一真相在 Go，由 internal/guard/parity_test.go
//     断言二者一致（漂移即失败）。
//
// 【只镜像常量，不镜像逻辑】：归一化（页码下界、页大小夹取）的唯一实现在 Go 的
// Request.Normalized()，本模块【不】照抄一份——那是把「一个规则两个实现」，
// 比复制常量更糟。前端发原始值，Go 归一化后经 Result 回填，前端以回显值为准。

/** 页大小上下界。单一真相在 Go 的 internal/page（DefaultPageSize / MaxPageSize），此处为镜像。 */
export const PageSize = {
  /** 请求未指定页大小（0）时 Go 采用的默认值。 */
  Default: 20,
  /** 页大小上限，超出由 Go 夹取。 */
  Max: 200,
} as const

/** 分页请求（对应 Go 的 page.Request）。字段名与 JSON tag 一致。 */
export interface PageRequest {
  /** 页码，1-based。小于 1 由 Go 归一化为 1。 */
  page: number
  /** 每页条数。0 由 Go 归一化为 PageSize.Default，超出上限被夹取。 */
  page_size: number
}

/** 分页结果（对应 Go 的 page.Result[T]）。 */
export interface PageResult<T> {
  /** 当前页数据。Go 侧保证为数组（无数据时是 []，不是 null）。 */
  list: T[]
  /** 满足条件的总条数。 */
  total: number
  /** Go 回填的归一化页码，前端应以它为准。 */
  page: number
  /** Go 回填的归一化页大小，前端应以它为准。 */
  page_size: number
}

/**
 * 构造分页请求（不归一化——归一化是 Go 的职责，见文件头）。
 * 省略参数时用 PageSize.Default 作为页大小。
 *
 * 参数显式标注 `number`：PageSize 用了 `as const`，默认值会被推断成字面量类型
 * （20 / 200），不标注则调用方传普通 number 时会类型不匹配。
 */
export function newPageRequest(page: number = 1, pageSize: number = PageSize.Default): PageRequest {
  return { page, page_size: pageSize }
}
